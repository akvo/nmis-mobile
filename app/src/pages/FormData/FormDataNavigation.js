import React from 'react';
import { View } from 'react-native';

const FormDataNavigation = () => {
  return <View />;
};

export default FormDataNavigation;
