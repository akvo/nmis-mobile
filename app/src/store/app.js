import { Store } from 'pullstate';

export const AppStore = new Store({
  currentPage: 'GetStarted',
  lang: 'en',
  online: false,
});
