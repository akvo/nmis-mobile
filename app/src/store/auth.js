import { Store } from 'pullstate';
/**
 * Server config
 */
export const AuthStore = new Store({
  authenticationTypes: ['assesment', 'username_password'], // list of authentication method
  authenticationType: 'assesment', // active type or current implementation of authentication type
  authenticationCode: '',
  serverUrl: 'http://api.example.com/nmis',
  username: '',
  password: '',
});
