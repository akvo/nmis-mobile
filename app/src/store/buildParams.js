import { Store } from 'pullstate';

export const BuildParamsStore = new Store({
  authenticationType: ['code_assignment', 'username', 'password'],
  serverURL: 'https://api.example.com/nmis',
  debugMode: false,
  dataSyncInterval: 300,
  errorHandling: true,
  loggingLevel: 'verbose',
  appVersion: '1.0.0',
});
