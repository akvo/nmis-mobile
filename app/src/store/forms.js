import { Store } from 'pullstate';

const FormState = new Store({
  form: {},
  questionGroups: [],
  questions: [],
  currentGroup: 1,
  saved: false,
  submitted: false,
  currentValues: {},
});

export default FormState;
