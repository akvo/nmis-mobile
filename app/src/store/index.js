export * from './app';
export * from './auth';
export * from './buildParams';
export * from './forms';
export * from './users';
