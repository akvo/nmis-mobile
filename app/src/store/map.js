import { Store } from 'pullstate';

const MapState = new Store({
  latitude: null,
  longitude: null,
});

export default MapState;
