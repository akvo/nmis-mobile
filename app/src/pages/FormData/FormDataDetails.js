import React, { useEffect, useState, useMemo, useCallback } from 'react';
import { View, Text, ScrollView, StyleSheet } from 'react-native';
import { ListItem } from '@rneui/themed';

import { FormState } from '../../store';
import { cascades } from '../../lib';
import { BaseLayout } from '../../components';
import FormDataNavigation from './FormDataNavigation';

const SubtitleContent = ({ index, answers, type, id, source }) => {
  const [cascadeValue, setCascadeValue] = useState(null);

  const fetchCascade = useCallback(async () => {
    if (source) {
      const cascadeID = parseInt(answers?.[id], 10);
      const { rows } = await cascades.loadDataSource(source, cascadeID);
      const { length: rowLength, _array: rowItems } = rows || {};
      const csValue = rowLength ? rowItems[0] : null;
      setCascadeValue(csValue);
    }
  }, [answers, id, source]);

  useEffect(() => {
    fetchCascade();
  }, [fetchCascade]);

  switch (type) {
    case 'geo':
      return (
        <View testID={`text-type-geo-${index}`}>
          <Text>Latitude: {answers?.[id]?.[0]}</Text>
          <Text>Longitude: {answers?.[id]?.[1]}</Text>
        </View>
      );

      break;
    case 'cascade':
      return <Text testID={`text-answer-${index}`}>{cascadeValue ? cascadeValue.name : '-'}</Text>;
      break;
    default:
      return <Text testID={`text-answer-${index}`}>{answers?.[id] || '-'}</Text>;
      break;
  }
};

const FormDataDetails = ({ navigation, route }) => {
  const selectedForm = FormState.useState((s) => s.form);
  const currentValues = FormState.useState((s) => s.currentValues);
  const [currentPage, setCurrentPage] = useState(0);

  const { json: formJSON } = selectedForm || {};

  const form = formJSON ? JSON.parse(formJSON) : {};
  const currentGroup = form?.question_group?.[currentPage] || [];
  const totalPage = form?.question_group?.length || 0;
  const questions = currentGroup?.question || [];

  useEffect(
    () =>
      navigation.addListener('beforeRemove', (e) => {
        // Prevent default behavior of leaving the screen
        e.preventDefault();

        if (Object.keys(currentValues).length) {
          FormState.update((s) => {
            s.currentValues = {};
          });
          navigation.dispatch(e.data.action);
        } else {
          return;
        }
      }),
    [navigation, currentValues],
  );

  return (
    <BaseLayout title={route?.params?.name} rightComponent={false}>
      <ScrollView>
        {questions?.map((q, i) => (
          <ListItem key={i} bottomDivider>
            <ListItem.Content>
              <ListItem.Title style={styles.title} testID={`text-question-${i}`}>
                {q.name}
              </ListItem.Title>
              <ListItem.Subtitle>
                <SubtitleContent index={i} answers={currentValues} {...q} />
              </ListItem.Subtitle>
            </ListItem.Content>
          </ListItem>
        ))}
      </ScrollView>
      <FormDataNavigation
        totalPage={totalPage}
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
      />
    </BaseLayout>
  );
};

const styles = StyleSheet.create({
  title: {
    fontWeight: 700,
    fontSize: 14,
    marginBottom: 4,
  },
});

export default FormDataDetails;
