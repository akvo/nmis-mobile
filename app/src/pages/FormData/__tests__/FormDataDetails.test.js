import React, { useState } from 'react';
import { render, renderHook, fireEvent, act, waitFor } from '@testing-library/react-native';
import { useNavigation } from '@react-navigation/native';

import { FormState } from '../../../store';
import FormDataNavigation from '../FormDataNavigation';
import FormDataDetails from '../FormDataDetails';
import { washInSchool, washInSchoolForm } from '../dummy-for-test-purpose';

const mockFormDataDetails = jest.fn();
jest.mock('../FormDataDetails', () => ({ formJSON, values, currentPage, setCurrentPage }) => {
  mockFormDataDetails(formJSON, values, currentPage, setCurrentPage);

  const form = formJSON ? JSON.parse(formJSON) : {};
  const currentGroup = form?.question_groups?.[currentPage] || [];
  const totalPage = form?.question_groups?.length || 0;
  const questions = currentGroup?.questions || [];
  const answers = values;

  return (
    <mock-View>
      {questions?.map((q, i) => (
        <mock-ListItem key={i} bottomDivider>
          <mock-ListItemContent>
            <mock-ListItemTitle testID={`text-question-${i}`}>{q.question}</mock-ListItemTitle>
            <mock-ListItemSubtitle>
              {q.type === 'geo' ? (
                <mock-View testID={`text-type-geo-${i}`}>
                  <mock-Text>Latitude: {answers?.[q.id]?.[0]}</mock-Text>
                  <mock-Text>Longitude: {answers?.[q.id]?.[1]}</mock-Text>
                </mock-View>
              ) : (
                <mock-Text testID={`text-answer-${i}`}>{answers?.[q.id] || q.id}</mock-Text>
              )}
            </mock-ListItemSubtitle>
          </mock-ListItemContent>
        </mock-ListItem>
      ))}

      <mock-FormDataNavigation>
        <mock-Text testID="text-pagination">
          {currentPage + 1}/{totalPage}
        </mock-Text>
      </mock-FormDataNavigation>
    </mock-View>
  );
});

describe('FormDataDetails', () => {
  beforeAll(() => {
    const { json: valuesJSON } = washInSchool;
    act(() => {
      FormState.update((s) => {
        s.form = {
          json: JSON.stringify(washInSchoolForm).replace(/'/g, "''"),
        };
        s.currentValues = JSON.parse(JSON.parse(valuesJSON));
      });
    });
  });

  it('should render correctly', () => {
    const { result: resultState } = renderHook(() => useState(0));
    const { result: resultForm } = renderHook(() => FormState.useState((s) => s.form));
    const { result: resultValues } = renderHook(() => FormState.useState((s) => s.currentValues));
    const [currentPage, setCurrentPage] = resultState.current;
    const { json: formJSON } = resultForm.current;
    const values = resultValues.current;

    const { getByTestId, debug } = render(
      <FormDataDetails
        formJSON={formJSON}
        values={values}
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
      />,
    );

    const questionText = getByTestId('text-question-0');
    expect(questionText).toBeDefined();
    const answerText = getByTestId('text-answer-0');
    expect(answerText).toBeDefined();
  });
  it('should list changed when navigation clicked', () => {
    const { result: resultState } = renderHook(() => useState(0));
    const { result: resultForm } = renderHook(() => FormState.useState((s) => s.form));
    const { result: resultValues } = renderHook(() => FormState.useState((s) => s.currentValues));
    const [currentPage, setCurrentPage] = resultState.current;
    const { json: formJSON } = resultForm.current;
    const values = resultValues.current;

    const { getByTestId, rerender, debug } = render(
      <FormDataDetails
        formJSON={formJSON}
        values={values}
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
      />,
    );

    act(() => {
      setCurrentPage(1);
    });

    rerender(
      <FormDataDetails
        formJSON={resultForm.current.json}
        values={resultValues.current}
        currentPage={resultState.current[0]}
        setCurrentPage={setCurrentPage}
      />,
    );

    const questionText = getByTestId('text-question-0');
    expect(questionText).toBeDefined();
    const answerText = getByTestId('text-answer-0');
    expect(answerText).toBeDefined();
    const paginationText = getByTestId('text-pagination');
    expect(paginationText).toBeDefined();
    expect(paginationText.props.children).toEqual([2, '/', 2]);
  });

  it('should match with snapshot', async () => {
    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        name: 'Datapoint name',
      },
    };

    const tree = render(<FormDataDetails navigation={mockNavigation} route={mockRoute} />);
    await waitFor(() => expect(tree.toJSON()).toMatchSnapshot());
  });
});
