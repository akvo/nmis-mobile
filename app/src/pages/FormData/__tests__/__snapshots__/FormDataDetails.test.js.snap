// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDataDetails should match with snapshot 1`] = `
<mock-View>
  <mock-FormDataNavigation>
    <mock-Text
      testID="text-pagination"
    >
      NaN
      /
      0
    </mock-Text>
  </mock-FormDataNavigation>
</mock-View>
`;
